File: back-end/core/src/test/java/fr/cph/crypto/core/TickerServiceImplTest.java
package fr.cph.crypto.core;

import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.util.Collections;
import java.util.List;

import fr.cph.crypto.core.api.entity.Currency;
import fr.cph.crypto.core.api.entity.Ticker;
import fr.cph.crypto.core.core.TickerServiceImpl;
import fr.cph.crypto.core.spi.TickerClient;
import fr.cph.crypto.core.spi.TickerRepository;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;

public class TickerServiceImplTest {

	private TickerClient tickerClient = Mockito.mock(TickerClient.class);
	private TickerRepository tickerRepository = Mockito.mock(TickerRepository.class);
	private TickerServiceImpl tickerService = new TickerServiceImpl(tickerClient, tickerRepository);

	@Test
	public void testFindOne() {
		// given
		String tickerId = "BTC-USD";
		Ticker ticker = new Ticker(null, Currency.BTC, Currency.USD, 1.0, "exchange", 0.0, 0.0, 0.0, 0.0, 0.0, 0L);
		given(tickerRepository.findOne(tickerId)).willReturn(ticker);

		// when
		Ticker actual = tickerService.findOne(tickerId);

		// then
		then(tickerRepository).should().findOne(tickerId);
		assertEquals(actual, ticker);
	}

	@Test
	public void testFindAllById() {
		// given
		List<String> tickerIds = Collections.singletonList("BTC-USD");
		Ticker ticker = new Ticker(null, Currency.BTC, Currency.USD, 1.0, "exchange", 0.0, 0.0, 0.0, 0.0, 0.0, 0L);
		given(tickerRepository.findAllById(tickerIds)).willReturn(Collections.singletonList(ticker));

		// when
		List<Ticker> actual = tickerService.findAllById(tickerIds);

		// then
		then(tickerRepository).should().findAllById(tickerIds);
		assertEquals(actual.get(0), ticker);
	}

	@Test
	public void testFindAll() {
		// given
		Ticker ticker = new Ticker(null, Currency.BTC, Currency.USD, 1.0, "exchange", 0.0, 0.0, 0.0, 0.0, 0.0, 0L);
		given(tickerRepository.findAll()).willReturn(Collections.singletonList(ticker));

		// when
		List<Ticker> actual = tickerService.findAll();

		// then
		then(tickerRepository).should().findAll();
		assertEquals(actual.get(0), ticker);
	}

	@Test
	public void testUpdateAll() {
		// given
		Ticker ticker = new Ticker(null, Currency.BTC, Currency.USD, 1.0, "exchange", 0.0, 0.0, 0.0, 0.0, 0.0, 0L);
		given(tickerClient.getTickers(Currency.USD, Currency.Companion.cryptoCurrenciesAsListOfString())).willReturn(Collections.singletonList(ticker));

		// when
		tickerService.updateAll();

		// then
		then(tickerClient).should().getTickers(Currency.USD, Currency.Companion.cryptoCurrenciesAsListOfString());
		then(tickerRepository).should().save(ticker);
	}
}

File: back-end/core/src/test/kotlin/fr/cph/crypto/core/TickerServiceImplTest.kt
package fr.cph.crypto.core

import fr.cph.crypto.core.api.entity.Currency
import fr.cph.crypto.core.api.entity.Ticker
import fr.cph.crypto.core.core.TickerServiceImpl
import fr.cph.crypto.core.spi.TickerClient
import fr.cph.crypto.core.spi.TickerRepository
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.BDDMockito.given
import org.mockito.BDDMockito.then
import org.mockito.Mockito

class TickerServiceImplTest {

    private val tickerClient = Mockito.mock(TickerClient::class.java)
    private val tickerRepository = Mockito.mock(TickerRepository::class.java)

    private val tickerService = TickerServiceImpl(tickerClient, tickerRepository)

    @Test
    fun testFindOne() {
        // given
        val tickerId = "BTC-USD"
        val ticker = Ticker(null, Currency.BTC, Currency.USD, 1.0, "exchange", 0.0, 0.0, 0.0, 0.0, 0.0, 0L)
        given(tickerRepository.findOne(tickerId)).willReturn(ticker)

        // when
        val actual = tickerService.findOne(tickerId)

        // then
        then(tickerRepository).should().findOne(tickerId)
        assertEquals(actual, ticker)
    }

    @Test
    fun testFindAllById() {
        // given
        val tickerIds = listOf("BTC-USD")
        val ticker = Ticker(null, Currency.BTC, Currency.USD, 1.0, "exchange", 0.0, 0.0, 0.0, 0.0, 0.0, 0L)
        given(tickerRepository.findAllById(tickerIds)).willReturn(listOf(ticker))

        // when
        val actual = tickerService.findAllById(tickerIds)

        // then
        then(tickerRepository).should().findAllById(tickerIds)
        assertEquals(actual[0], ticker)
    }

    @Test
    fun testFindAll() {
        // given
        val ticker = Ticker(null, Currency.BTC, Currency.USD, 1.0, "exchange", 0.0, 0.0, 0.0, 0.0, 0.0, 0L)
        given(tickerRepository.findAll()).willReturn(listOf(ticker))

        // when
        val actual = tickerService.findAll()

        // then
        then(tickerRepository).should().findAll()
        assertEquals(actual[0], ticker)
    }

    @Test
    fun testUpdateAll() {
        // given
        val ticker = Ticker(null, Currency.BTC, Currency.USD, 1.0, "exchange", 0.0, 0.0, 0.0, 0.0, 0.0, 0L)
        given(tickerClient.getTickers(Currency.USD, Currency.cryptoCurrenciesAsListOfString())).willReturn(listOf(ticker))

        // when
        tickerService.updateAll()

        // then
        then(tickerClient).should().getTickers(Currency.USD, Currency.cryptoCurrenciesAsListOfString())
        then(tickerRepository).should().save(ticker)
    }
}